core/storage/question/gae_models.py

```diff
@@ -0,0 +1,113 @@
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS-IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Models for storing the question data models."""
+
+from core.platform import models
+
+from google.appengine.ext import ndb
+
+import utils
+
+(base_models,) = models.Registry.import_models([models.NAMES.base_model])
+
+class QuestionSnapshotMetadataModel(base_models.BaseSnapshotMetadataModel):
+    """Storage model for the metadata for a question snapshot."""
+    pass
+
+
+class QuestionSnapshotContentModel(base_models.BaseSnapshotContentModel):
+    """Storage model for the content of a question snapshot."""
+    pass
+
+
+class QuestionModel(base_models.VersionedModel):
+    """Model for storing Questions.
+
+    The id of instances of this class has the form
+    {{collection_id}}.{{random_hash_of_16_chars}}
+    """
+    SNAPSHOT_METADATA_CLASS = QuestionSnapshotMetadataModel
+    SNAPSHOT_CONTENT_CLASS = QuestionSnapshotContentModel
+    ALLOW_REVERT = True
+
+    # The title of the question.
+    title = ndb.StringProperty(required=True, indexed=True)
+    # The data of the question.
+    question_data = ndb.JsonProperty(default={}, indexed=False)
+    # The schema version for the data.
+    data_schema_version = ndb.IntegerProperty(required=True, indexed=True)
+    # The ID of collection to which the question belongs.
+    collection_id = ndb.IntegerProperty(required=True, indexed=True)
+    # The ISO 639-1 code for the language this question is written in.
+    language_code = ndb.StringProperty(required=True, indexed=True)
+
+    @classmethod
+    def _get_new_id(cls, collection_id):
+        """Generates a unique id for the training job of the form
+        {{collection_id}}.{{random_hash_of_16_chars}}
+
+        Args:
+            collection_id: int. ID of the collection.
+
+        Returns:
+            ID of the new QuestionModel instance.
+
+        Raises:
+            Exception: The id generator for QuestionModel is
+            producing too many collisions.
+        """
+
+        for _ in range(base_models.MAX_RETRIES):
+            new_id = '%s.%s' % (
+                collection_id,
+                utils.convert_to_hash(
+                    str(utils.get_random_int(base_models.RAND_RANGE)),
+                    base_models.ID_LENGTH))
+            if not cls.get_by_id(new_id):
+                return new_id
+
+        raise Exception(
+            'The id generator for QuestionModel is producing too many '
+            'collisions.')
+
+    @classmethod
```

Comment:
Tests?

Reply:
I am sorry. Does this mean to write tests to check for identical ids?

Reply:
I meant writing tests for this method in gae_models_test -- e.g. create a storage model, retrieve it, and check that the fields are set correctly.

Reply:
I think it is being tested in question_services_test using test_save_new_question method.
Wouldn't it be redundant to do that again in age_models_test?

Reply:
No; gae_models_test should test the functions in the storage layer, and question_services_test should test the functions in the question_services file. The purposes are different.